File: techDose/Graph/division.js
class UnionFind {
    constructor() {
        this.parent = {}; // Stores the parent of each node
        this.weight = {}; // Stores the weight of each node relative to its parent
    }
    debugger;

    // Finds the root of the given node while applying path compression
    find(node) {
        if (!(node in this.parent)) {
            this.parent[node] = node; // Initialize parent of the node
            this.weight[node] = 1.0; // Initialize weight of the node
            return node;
        }

        if (this.parent[node] !== node) {
            const originalParent = this.parent[node];
            this.parent[node] = this.find(originalParent); // Path compression
            this.weight[node] *= this.weight[originalParent]; // Update the weight relative to the root
        }

        return this.parent[node];
    }

    // Unites two nodes with a given ratio
    union(node1, node2, value) {
        const root1 = this.find(node1);
        const root2 = this.find(node2);

        if (root1 !== root2) {
            this.parent[root1] = root2;
            this.weight[root1] = (value * this.weight[node2]) / this.weight[node1];
        }
    }

    // Checks if two nodes are connected and returns the division result if they are
    isConnected(node1, node2) {
        if (!(node1 in this.parent) || !(node2 in this.parent)) {
            return -1.0; // One of the nodes is not in the graph
        }

        const root1 = this.find(node1);
        const root2 = this.find(node2);

        if (root1 !== root2) {
            return -1.0; // Nodes are not connected
        }

        return this.weight[node1] / this.weight[node2];
    }
}

function calcEquation(equations, values, queries) {
    const uf = new UnionFind();

    // Build the union-find structure
    for (let i = 0; i < equations.length; i++) {
        const [var1, var2] = equations[i];
        uf.union(var1, var2, values[i]);
    }

    // Process the queries
    const results = [];
    for (const [var1, var2] of queries) {
        results.push(uf.isConnected(var1, var2));
    }

    return results;
}

// Example Input
const equations = [["a","b"],["b","c"]];
const values = [2.0,3.0];
const queries = [["a","c"],["b","a"],["a","e"],["a","a"],["x","x"]];

// Output the results
debugger;
console.log(calcEquation(equations, values, queries));

File: techDose/Graph/division.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <string>
using namespace std;

class UnionFind {
private:
    unordered_map<string, string> parent; // Parent of each node
    unordered_map<string, double> weight; // Weight of each node relative to its parent

public:
    // Find the root of the node with path compression
    string find(const string& node) {
        if (parent.find(node) == parent.end()) {
            this->parent[node] = node;
            weight[node] = 1.0;
            return node;
        }

        if (parent[node] != node) {
            string originalParent = parent[node];
            parent[node] = find(originalParent);
            weight[node] *= weight[originalParent];
        }
        return parent[node];
    }

    // Unite two nodes with a given ratio
    void unionNodes(const string& node1, const string& node2, double value) {
        string root1 = find(node1);
        string root2 = find(node2);

        if (root1 != root2) {
            parent[root1    ] = root2;
            weight[root1] = (value * weight[node2]) / weight[node1];
        }
    }

    // Check if two nodes are connected and return their division result if they are
    double isConnected(const string& node1, const string& node2) {
        if (parent.find(node1) == parent.end() || parent.find(node2) == parent.end()) {
            return -1.0; // One of the nodes is not in the graph
        }

        string root1 = find(node1);
        string root2 = find(node2);

        if (root1 != root2) {
            return -1.0; // Nodes are not connected
        }

        return weight[node1] / weight[node2];
    }
};

vector<double> calcEquation(
    vector<pair<string, string>>& equations, 
    vector<double>& values, 
    vector<pair<string, string>>& queries
) {
    UnionFind uf;

    // Build the union-find structure
    for (size_t i = 0; i < equations.size(); ++i) {
        const string& var1 = equations[i].first;
        const string& var2 = equations[i].second;
        uf.unionNodes(var1, var2, values[i]);
    }

    // Process the queries
    vector<double> results;
    for (const auto& query : queries) {
        results.push_back(uf.isConnected(query.first, query.second));
    }

    return results;
}

int main() {
    vector<pair<string, string>> equations = {{"a", "b"}, {"b", "c"}};
    vector<double> values = {2.0, 3.0};
    vector<pair<string, string>> queries = {{"a", "c"}, {"b", "a"}};

    vector<double> results = calcEquation(equations, values, queries);

    for (double result : results) {
        cout << result << endl;
    }

    return 0;
}
